File: go/otp.go
package callisto

import (
	"context"
	"encoding/json"
)

// OTPService accesses the OTP resource.
type OTPService struct {
	t *transport
}

// SendOTPParams are the parameters for OTP.Send.
type SendOTPParams struct {
	// To is the recipient. Required.
	To string
	// Message is the message template/body. Required.
	Message string
	// Sender is the optional sender name.
	Sender string
	// ExpiredIn is the optional code lifetime in seconds.
	ExpiredIn *int
	// Type is the optional OTP character set.
	Type OtpType
	// DigitSize is the optional code length.
	DigitSize *int
	// Provider is the optional delivery channel (sms or whatsapp).
	Provider OtpProvider
	// InstanceCode is the WhatsApp instance code. Required when Provider is
	// whatsapp (sent as the JSON key "instanceCode").
	InstanceCode string
}

// Send generates and sends an OTP. POST /otp/send.
//
// Client-side validation: when Provider is whatsapp, InstanceCode is required;
// otherwise a *ValidationError is returned before any request is made.
func (s *OTPService) Send(ctx context.Context, params SendOTPParams) (*SendOTPResult, error) {
	if params.Provider == OtpProviderWhatsApp && params.InstanceCode == "" {
		verr := &ValidationError{CallistoError: &CallistoError{
			Message: "instance_code is required when provider is whatsapp",
		}}
		s.t.reportLocal(verr)
		return nil, verr
	}

	body := map[string]any{"to": params.To, "message": params.Message}
	if params.Sender != "" {
		body["sender"] = params.Sender
	}
	if params.ExpiredIn != nil {
		body["expired_in"] = *params.ExpiredIn
	}
	if params.Type != "" {
		body["type"] = string(params.Type)
	}
	if params.DigitSize != nil {
		body["digit_size"] = *params.DigitSize
	}
	if params.Provider != "" {
		body["provider"] = string(params.Provider)
	}
	if params.InstanceCode != "" {
		body["instanceCode"] = params.InstanceCode
	}

	raw, err := s.t.request(ctx, "POST", "/otp/send", body, nil)
	if err != nil {
		return nil, err
	}
	var out SendOTPResult
	if err := json.Unmarshal(raw, &out); err != nil {
		return nil, err
	}
	return &out, nil
}

// Verify checks a submitted OTP code. POST /otp/verify.
func (s *OTPService) Verify(ctx context.Context, otpID, code string) (*VerifyOTPResult, error) {
	body := map[string]any{"otp_id": otpID, "code": code}
	raw, err := s.t.request(ctx, "POST", "/otp/verify", body, nil)
	if err != nil {
		return nil, err
	}
	var out VerifyOTPResult
	if err := json.Unmarshal(raw, &out); err != nil {
		return nil, err
	}
	return &out, nil
}

// GetStatus fetches a single OTP by ID. GET /otps/{id}.
func (s *OTPService) GetStatus(ctx context.Context, otpID string) (*OTP, error) {
	raw, err := s.t.request(ctx, "GET", "/otps/"+otpID, nil, nil)
	if err != nil {
		return nil, err
	}
	var out OTP
	if err := json.Unmarshal(raw, &out); err != nil {
		return nil, err
	}
	return &out, nil
}

// ListOTPParams are the parameters for OTP.List.
type ListOTPParams struct {
	StartedAt string
	EndedAt   string
	Page      *int
	Limit     *int
}

// List lists OTP records. GET /otps.
func (s *OTPService) List(ctx context.Context, params ListOTPParams) (*Paginated[OTP], error) {
	query := map[string]any{
		"started_at": emptyToNil(params.StartedAt),
		"ended_at":   emptyToNil(params.EndedAt),
		"page":       params.Page,
		"limit":      params.Limit,
	}
	raw, err := s.t.request(ctx, "GET", "/otps", nil, query)
	if err != nil {
		return nil, err
	}
	var out Paginated[OTP]
	if err := json.Unmarshal(raw, &out); err != nil {
		return nil, err
	}
	return &out, nil
}

File: ruby/lib/callisto/resources/otp.rb
# frozen_string_literal: true

module Callisto
  # OTP resource.
  class OtpResource
    # @param transport [Transport]
    def initialize(transport)
      @t = transport
    end

    # Generates and sends a one-time password.
    #
    # @param to [String] recipient number
    # @param message [String] message template
    # @param sender [String, nil] sender name
    # @param expired_in [Integer, nil] code lifetime in seconds
    # @param type [String, nil] OtpType value or raw string
    # @param digit_size [Integer, nil] number of characters in the code
    # @param provider [String, nil] OtpProvider value or raw string
    # @param instance_code [String, nil] WhatsApp instance code (required when provider is whatsapp)
    # @return [SendOtpResult]
    # @raise [ValidationError] when provider is "whatsapp" and instance_code is missing
    def send(to:, message:, sender: nil, expired_in: nil, type: nil,
             digit_size: nil, provider: nil, instance_code: nil)
      if provider == OtpProvider::WHATSAPP && (instance_code.nil? || instance_code.empty?)
        err = ValidationError.new("instance_code is required when provider is whatsapp")
        @t.reporter&.capture_exception(err) if @t.respond_to?(:reporter)
        raise err
      end

      body = { "to" => to, "message" => message }
      body["sender"] = sender unless sender.nil?
      body["expired_in"] = expired_in unless expired_in.nil?
      body["type"] = type unless type.nil?
      body["digit_size"] = digit_size unless digit_size.nil?
      body["provider"] = provider unless provider.nil?
      body["instanceCode"] = instance_code unless instance_code.nil?

      SendOtpResult.from_hash(@t.request("POST", "/otp/send", body: body))
    end

    # Verifies a code against an OTP.
    #
    # @param otp_id [String]
    # @param code [String]
    # @return [VerifyOtpResult]
    def verify(otp_id:, code:)
      VerifyOtpResult.from_hash(
        @t.request("POST", "/otp/verify", body: { "otp_id" => otp_id, "code" => code })
      )
    end

    # Fetches a single OTP by ID.
    #
    # @param otp_id [String]
    # @return [Otp]
    def get_status(otp_id)
      Otp.from_hash(@t.request("GET", "/otps/#{otp_id}"))
    end

    # Lists OTPs.
    #
    # @param started_at [String, nil]
    # @param ended_at [String, nil]
    # @param page [Integer, nil]
    # @param limit [Integer, nil]
    # @return [Paginated] of Otp
    def list(started_at: nil, ended_at: nil, page: nil, limit: nil)
      data = @t.request("GET", "/otps", query: {
        "started_at" => started_at, "ended_at" => ended_at,
        "page" => page, "limit" => limit
      })
      Paginated.from_hash(data, Otp.method(:from_hash))
    end
  end
end
